### DISCUSSION TITLE
Czech Keyboard

### DISCUSSION THREAD

**Discussion by dmandinec** [neutrinolabs/xrdp#2584] [Q&A]
So i have this problem. I am currently using Fedora 37 and kali linux with xrdp. But i have this problem. I am from czech republic and I am using Czech keyboard layout. But the catch is in xrdp keyboard layouts. I have not found any in local files and neither here on github. So my question is. Can i find that keyboard layout here or i am screwd?
ps. the name of keyboard layout should be "km-00000405.ini"

**Comment by matt335672** [comment]
Take a look at neutrinolabs/xrdp#2337.

The `xrdp-genkeymap` command should let you generate a keymap.

**Reply by dmandinec** [reply]
it did not help, i followed discusion but still, there is US keyboard. So is it unchangable or it can be changed?
the `xrdp-genkeymap` generates it, i think so, but still it will not work.

**Reply by matt335672** [reply]
Generate your keyboard layout, then check it's being loaded in `xrdp.log` when the client connects. You may have a problem loading it, in which case the system will fall back to US. For example, file permissions may be causing problems.

**Comment by rammicz** [comment]
Had the same issue on Ubuntu 22.04 / xrdp 0.9.17. Layout kept reverting to US
  English on every RDP reconnect. The root cause is that Czech layout 0x00000405
   is completely missing from both xrdp_keyboard.ini and the km-*.ini keymap
  files.

  Fix:

  1. Generate the Czech keymap file (make sure cz layout is active first):
  setxkbmap cz
  xrdp-genkeymap /etc/xrdp/km-00000405.ini
  2. Add Czech to /etc/xrdp/xrdp_keyboard.ini:
    - In [default_rdp_layouts] section add: rdp_layout_cz=0x00000405
    - In [default_layouts_map] section add: rdp_layout_cz=cz
  3. Optionally set XkbLayout "cz" in /etc/X11/xrdp/xorg.conf keyboard section
  4. Restart xrdp: sudo systemctl restart xrdp

  See also #3348 which has an official patch for this on v0.9.x.

**Reply by matt335672** [reply]
Thanks @rammicz 

If you're at all able, I'd suggest an upgrade to v0.9.27 (or later), due to [CVE-2025-68670: Improper bounds checking of domain string length leads to Stack-based Buffer Overflow](https://nvd.nist.gov/vuln/detail/CVE-2025-68670). If you're not able to upgrade, please make yourself aware of the contents of the CVE, so you can decide whether or not it applies to your environment.

**Reply by matt335672** [reply]
Just edited the above - you need v0.9.27, not v0.9.26.
